cpp/src/parquet/encoding.cc

```diff
@@ -1238,43 +1240,93 @@ int PlainBooleanDecoder::Decode(bool* buffer, int max_values) {
   return max_values;
 }
 
-struct ArrowBinaryHelper {
-  explicit ArrowBinaryHelper(typename EncodingTraits<ByteArrayType>::Accumulator* out) {
-    this->out = out;
-    this->builder = out->builder.get();
-    this->chunk_space_remaining =
-        ::arrow::kBinaryMemoryLimit - this->builder->value_data_length();
+template <typename DType, typename Enable = void>
+struct ArrowBinaryHelper;
+
+template <typename DType>
+struct ArrowBinaryHelper<DType,
+                         std::enable_if_t<std::is_same_v<DType, ByteArrayType>, void>> {
+  explicit ArrowBinaryHelper(typename EncodingTraits<DType>::Accumulator* acc) {
+    builder = acc->builder.get();
+    chunks = &acc->chunks;
+    if (ARROW_PREDICT_FALSE(SubtractWithOverflow(::arrow::kBinaryMemoryLimit,
+                                                 builder->value_data_length(),
+                                                 &chunk_space_remaining))) {
+      throw ParquetException("excess expansion in ArrowBinaryHelper<DType>");
+    }
   }
 
   Status PushChunk() {
     std::shared_ptr<::arrow::Array> result;
     RETURN_NOT_OK(builder->Finish(&result));
-    out->chunks.push_back(result);
+    chunks->push_back(result);
     chunk_space_remaining = ::arrow::kBinaryMemoryLimit;
     return Status::OK();
   }
 
   bool CanFit(int64_t length) const { return length <= chunk_space_remaining; }
 
   void UnsafeAppend(const uint8_t* data, int32_t length) {
+    DCHECK(CanFit(length));
     chunk_space_remaining -= length;
     builder->UnsafeAppend(data, length);
   }
 
   void UnsafeAppendNull() { builder->UnsafeAppendNull(); }
 
   Status Append(const uint8_t* data, int32_t length) {
+    DCHECK(CanFit(length));
     chunk_space_remaining -= length;
     return builder->Append(data, length);
   }
 
   Status AppendNull() { return builder->AppendNull(); }
 
-  typename EncodingTraits<ByteArrayType>::Accumulator* out;
-  ::arrow::BinaryBuilder* builder;
+  typename EncodingTraits<DType>::BuilderType* builder;
+  std::vector<std::shared_ptr<::arrow::Array>>* chunks;
   int64_t chunk_space_remaining;
 };
 
+template <typename DType>
+struct ArrowBinaryHelper<DType, std::enable_if_t<std::is_same_v<DType, FLBAType>, void>> {
+  explicit ArrowBinaryHelper(typename EncodingTraits<DType>::Accumulator* acc) {
+    builder = acc;
+    if (ARROW_PREDICT_FALSE(SubtractWithOverflow(::arrow::kBinaryMemoryLimit,
+                                                 builder->value_data_length(),
+                                                 &space_remaining))) {
+      throw ParquetException("excess expansion in ArrowBinaryHelper<DType>");
+    }
+  }
+
+  Status PushChunk() {
+    std::shared_ptr<::arrow::Array> result;
+    RETURN_NOT_OK(builder->Finish(&result));
```

Comment:
I'm reading through this again and I don't understand what you're doing. Are you throwing away the `result`?

Reply:
Indeed this would not make sense. I've pushed a change for `ArrowBinaryHelper<FLBAType>` to have an internal `std::vector<std::shared_ptr<::arrow::Array>>* chunks` and store results there.

Reply:
See https://github.com/apache/arrow/pull/14341/commits/8f4a54adec0ef4ba7e2d3cc4cae1b5098a172034.